src/rust/uid-allocator/src/counters_db.rs

```diff
@@ -0,0 +1,54 @@
+use sqlx::PgPool;
+
+use crate::{
+    allocator::PreAllocation,
+    service::UidAllocatorServiceError,
+};
+
+#[derive(Debug)]
+pub struct Count {
+    pub prev: i64,
+    pub new: i64,
+}
+
+#[derive(Clone)]
+pub struct CountersDb {
+    pub pool: PgPool,
+}
+
+impl CountersDb {
+    pub async fn preallocate(
+        &self,
+        tenant_id: uuid::Uuid,
+        size: u32,
+    ) -> Result<PreAllocation, UidAllocatorServiceError> {
+        let mut conn = self.pool.acquire().await?;
+        // Increments the tenant's allocation counter by `size`, and returns the new value as well as the previous value
+        let count = sqlx::query_as!(
+            Count,
+            "
+            UPDATE counters
+            SET counter = counter + $1
+            FROM (
+                     SELECT counter as prev
+                     FROM counters
+                     WHERE counters.tenant_id = $2
+                     LIMIT 1
+                 ) as c
+            WHERE counters.tenant_id = $2
+            RETURNING counter as new, c.prev
+            ",
+            size as i32,
+            tenant_id
```

Comment:
Just to make sure, this is multiprocess-safe? I don't really know postgres's lock guarantees

Reply:
I should add FOR UPDATE to the select just in case.